vpp-manager/uplink/rdma.go

```diff
@@ -0,0 +1,139 @@
+// limitations under the License.
+
+package uplink
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/projectcalico/vpp-dataplane/vpp-manager/config"
+	"github.com/projectcalico/vpp-dataplane/vpp-manager/utils"
+	"github.com/projectcalico/vpp-dataplane/vpplink"
+	log "github.com/sirupsen/logrus"
+)
+
+type RDMADriver struct {
+	UplinkDriverData
+}
+
+func (d *RDMADriver) IsSupported(warn bool) (supported bool) {
+	var ret bool
+	supported = true
+	ret = d.params.LoadedDrivers[config.DRIVER_VFIO_PCI]
+	if !ret && warn {
+		log.Warnf("did not find vfio-pci or uio_pci_generic driver")
+		log.Warnf("VPP may fail to grab its interface")
+	}
+	supported = supported && ret
+
+	ret = d.params.AvailableHugePages > 0
+	if !ret && warn {
+		log.Warnf("No hugepages configured, driver won't work")
+	}
+	supported = supported && ret
+
+	ret = d.conf.Driver == config.DRIVER_VIRTIO_PCI
+	if !ret && warn {
+		log.Warnf("Interface driver is <%s>, not %s", d.conf.Driver, config.DRIVER_VIRTIO_PCI)
+	}
+	supported = supported && ret
+
+	return supported
+}
+
+func (d *RDMADriver) PreconfigureLinux() (err error) {
+	newDriverName := d.params.NewDriverName
+	doSwapDriver := d.conf.DoSwapDriver
+	/*if newDriverName == "" {
+		newDriverName = config.DRIVER_VFIO_PCI
+		doSwapDriver = config.DRIVER_VFIO_PCI != d.conf.Driver
+	}*/
+
+	/*if !d.params.VfioUnsafeiommu {
+		err := utils.SetVfioUnsafeiommu(true)
+		if err != nil {
+			return errors.Wrapf(err, "Virtio preconfigure error")
+		}
+	}*/
+	d.removeLinuxIfConf(true /* down */)
+	if doSwapDriver {
+		err = utils.SwapDriver(d.conf.PciId, newDriverName, true)
+		if err != nil {
+			log.Warnf("Failed to swap driver to %s: %v", newDriverName, err)
+		}
+	}
+	return nil
+}
+
+func (d *RDMADriver) RestoreLinux() {
+	if !d.params.VfioUnsafeiommu {
+		err := utils.SetVfioUnsafeiommu(false)
+		if err != nil {
+			log.Warnf("Virtio restore error %v", err)
+		}
+	}
+	if d.conf.PciId != "" && d.conf.Driver != "" {
+		err := utils.SwapDriver(d.conf.PciId, d.conf.Driver, false)
+		if err != nil {
+			log.Warnf("Error swapping back driver to %s for %s: %v", d.conf.Driver, d.conf.PciId, err)
+		}
+	}
+	if !d.conf.IsUp {
+		return
+	}
+	// This assumes the link has kept the same name after the rebind.
+	// It should be always true on systemd based distros
+	link, err := utils.SafeSetInterfaceUpByName(d.params.MainInterface)
+	if err != nil {
+		log.Warnf("Error setting %s up: %v", d.params.MainInterface, err)
+		return
+	}
+
+	// Re-add all adresses and routes
+	d.restoreLinuxIfConf(link)
+}
+
+func (d *RDMADriver) CreateMainVppInterface(vpp *vpplink.VppLink, vppPid int) (err error) {
+	MyInfo := vpplink.RDMAInfo{
+	d.params.MainInterface,
+	d.params.NumRxQueues,
+	d.params.RxQueueSize,
+	d.params.TxQueueSize,
+	}
+	swIfIndex, err := vpp.CreateRDMA(MyInfo)
+
+	if err != nil {
+		return errors.Wrapf(err, "Error creating RDMA interface")
+	}
+	err = d.moveInterfaceToNS(d.params.MainInterface, vppPid)
```

Comment:
Please check the value of err here and return an error if moving the interface fails.